pydantic/functional_validators.py

```diff
@@ -153,13 +153,17 @@ class Model(BaseModel):
     func: core_schema.NoInfoValidatorFunction | core_schema.WithInfoValidatorFunction
 
     def __get_pydantic_core_schema__(self, source_type: Any, handler: _GetCoreSchemaHandler) -> core_schema.CoreSchema:
+        schema = handler(source_type)
+        serialization = core_schema.wrap_serializer_function_ser_schema(function=lambda v, h: h(v), schema=schema)
```

Comment:
I think we can avoid introducing a wrapper here and just do

```suggestion
        serialization = schema.get('serialization')
```

Reply:
Hello.

Thanks for the advice. I originally tried that, but ended up with some type checking errors:
```
Typecheck................................................................Failed
- hook id: typecheck
- exit code: 1

/home/anvil/git/pydantic/pydantic/functional_validators.py
  /home/anvil/git/pydantic/pydantic/functional_validators.py:163:68 - error: Impossible d’affecter l’argument de type « SerSchema | IncExDictOrElseSerSchema | IncExSeqOrElseSerSchema | None » au paramètre « serialization » de type « SerSchema | None » dans la fonction « with_info_plain_validator_function » (reportGeneralTypeIssues)
  /home/anvil/git/pydantic/pydantic/functional_validators.py:167:85 - error: Impossible d’affecter l’argument de type « SerSchema | IncExDictOrElseSerSchema | IncExSeqOrElseSerSchema | None » au paramètre « serialization » de type « SerSchema | None » dans la fonction « no_info_plain_validator_function » (reportGeneralTypeIssues)
2 errors, 0 warnings, 0 informations
```

Should I cast the value ?

Reply:
ha, pardon my french.

```
PDM, version 2.11.2
pdm run ruff pydantic tests docs/plugins
pdm run ruff format --check pydantic tests docs/plugins
164 files already formatted
pre-commit 3.6.0
pre-commit run typecheck --all-files
Typecheck................................................................Failed
- hook id: typecheck
- exit code: 1

/home/anvil/git/pydantic/pydantic/functional_validators.py
  /home/anvil/git/pydantic/pydantic/functional_validators.py:163:68 - error: Argument of type "SerSchema | IncExDictOrElseSerSchema | IncExSeqOrElseSerSchema | None" cannot be assigned to parameter "serialization" of type "SerSchema | None" in function "with_info_plain_validator_function" (reportGeneralTypeIssues)
  /home/anvil/git/pydantic/pydantic/functional_validators.py:167:85 - error: Argument of type "SerSchema | IncExDictOrElseSerSchema | IncExSeqOrElseSerSchema | None" cannot be assigned to parameter "serialization" of type "SerSchema | None" in function "no_info_plain_validator_function" (reportGeneralTypeIssues)
2 errors, 0 warnings, 0 informations
```

Reply:
Potentially instead of the `assert` we can fall back to the slower wrap serializer proposed here just to be safe?

Reply:
Maybe. Can we first investigate if the right answer is to add those serializers to the union in core?

Reply:
e.g. `if serializiation['type'] in (...)  # slow paths else # fast paths`

Reply:
@davidhewitt @adriangb, we might as well investigate adding them to the union in core, as we'll have to do a 2.16.1 patch to core before we do the 2.6 pydantic release

Reply:
I think it's worth at least looking into be cause that might be a bug / omission!

Reply:
I dont get all the details (yet?), but if I can be of any use, let me know.

Reply:
Ah, adding the inc/exc stuff to the `SerSchema` union isn't easily done - for ex:
```
IncExSeqOrElseSerSchema = Union[IncExSeqSerSchema, SerSchema]
IncExDictOrElseSerSchema = Union[IncExDictSerSchema, SerSchema]
```

We would end up having recursive defs in the `SerSchema` union, as this is how those types are defined. For now, I think it makes sense to do a type check, and perhaps we could add an issue to the `pydantic-core` board if we want to figure out a way to add those types to the `SerSchema` Union in the long term.

Reply:
Alright, a proposed fix:

```py
schema = handler(source_type)

        # if the serialization schema is of type 'include-exclude-dict' or 'include-exclude-set',
        # we need to wrap it in a serializer function schema
        # otherwise, we can leave it as is (which is more efficient)
        serialization = schema.get('serialization')
        if serialization and serialization['type'] in ('include-exclude-dict', 'include-exclude-set'):
            serialization = core_schema.wrap_serializer_function_ser_schema(function=lambda v, h: h(v), schema=schema)

        info_arg = _inspect_validator(self.func, 'plain')
        if info_arg:
            func = cast(core_schema.WithInfoValidatorFunction, self.func)
            return core_schema.with_info_plain_validator_function(
                func,
                field_name=handler.field_name,
                serialization=serialization,  # type: ignore
            )
        else:
            func = cast(core_schema.NoInfoValidatorFunction, self.func)
            return core_schema.no_info_plain_validator_function(func, serialization=serialization)  # type: ignore
```

This requires some `# type: ignore` statements, which isn't ideal. However, it's not possible to check if something is an instance of `SerSchema`, as that's a union of types. Furthermore, checking against the individual types is difficult bc they're often `TypedDict` types, which aren't friendly with isinstance checks.

Reply:
Actually, let's go with:

```py
schema = handler(source_type)

_ser_schema_serialization_types = (
    *get_args(core_schema.ExpectedSerializationTypes),
    'function-plain',
    'function-wrap',
    'format',
    'to-string',
    'function',
    'model',
)

# if the serialization schema is of type 'include-exclude-dict' or 'include-exclude-set',
# we need to wrap it in a serializer function schema
# otherwise, we can leave it as is (which is more efficient)
serialization = schema.get('serialization')
if serialization:
    if serialization['type'] in ('include-exclude-dict', 'include-exclude-set'):
        serialization = core_schema.wrap_serializer_function_ser_schema(
            function=lambda v, h: h(v), schema=schema
        )
    else:
        assert serialization['type'] in _ser_schema_serialization_types

info_arg = _inspect_validator(self.func, 'plain')
if info_arg:
    func = cast(core_schema.WithInfoValidatorFunction, self.func)
    return core_schema.with_info_plain_validator_function(
        func,
        field_name=handler.field_name,
        serialization=serialization,  # type: ignore
    )
else:
    func = cast(core_schema.NoInfoValidatorFunction, self.func)
    return core_schema.no_info_plain_validator_function(func, serialization=serialization)  # type: ignore
```

Or something along those lines. @Anvil, could you please implement that change + we can get this across the line today?

Just as a side note - this approach still feels a bit unclean to me, so I'm going to mark this as something to come back to and see if we can clean up.

Reply:
Alright, after discussing a bit more with @adriangb, I think that my above suggestion is ugly enough that we shouldn't use that alternative.

@Anvil, let's go with what you currently have, and I'll create an issue to address the performance of the 
```
serialization = core_schema.wrap_serializer_function_ser_schema(function=lambda v, h: h(v), schema=schema)
```

line that you've added. I'd rather have the bug fix now and improve the performance down the line.

Thanks for all of your help on this!